### PR TITLE
Support lighthouse reports

### PR DESCRIPTION
**Pull request by kqmpetenz**
added the possibility to create lighthouse reports

### CODE REVIEW

File: src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+     * Creates a <a href="https://developer.chrome.com/docs/lighthouse/overview?hl=de">Lighthouse</a> report
+     * (Copyright Google) of the current URL and adds it to the Allure report.
+     * 
+     * @param driver
+     *            The current webdriver
+     * @param reportName
+     *            The name of the Lighthouse report attachment in the Allure report
+     * @throws Exception
+     */
+    public static void createLightHouseReport(WebDriver driver, String reportName) throws Exception
```
Comment (wurzelkuchen) [inline src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java:38]: We should make the interface as simple as possible. There is no need to have the webdriver given as a parameter. Just use `Neodymium.getWebdriver()` inside this method.

File: src/test/java/com/xceptance/neodymium/util/LighthouseUtilsTest.java [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        
+        if (System.getProperty("os.name").toLowerCase().contains("win")) 
+        {
+            Neodymium.configuration().setProperty("neodymium.lighthouse.binaryPath", "echo {\"categories\": {\"performance\": {\"score\": 0.5}, \"accessibility\": {\"score\": 0.5}, \"best-practices\": {\"score\": 0.5}, \"seo\": {\"score\": 0.5}}} > target/lighthouseUtilsReport.report.json | echo makeCommentWork #");
+            builder = new ProcessBuilder("cmd.exe", "/c", "fabricatedHtml > target/lighthouseUtilsReport.report.html");
+        }
+        else if (System.getProperty("os.name").toLowerCase().contains("linux") || System.getProperty("os.name").toLowerCase().contains("mac"))
+        {
+            Neodymium.configuration().setProperty("neodymium.lighthouse.binaryPath", "echo {\"categories\": {\"performance\": {\"score\": 0.5}, \"accessibility\": {\"score\": 0.5}, \"best-practices\": {\"score\": 0.5}, \"seo\": {\"score\": 0.5}}} > target/lighthouseUtilsReport.report.json");
+            builder = new ProcessBuilder("fabricatedHtml > target/lighthouseUtilsReport.report.html");
```
Comment (wurzelkuchen) [inline src/test/java/com/xceptance/neodymium/util/LighthouseUtilsTest.java:33]: this needs to be changed to `new ProcessBuilder("sh", "-c", "echo fabricatedHtml > target/lighthouseUtilsReport.report.html");`

File: src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    {
+        ProcessBuilder builder = new ProcessBuilder();
+        
+        if (System.getProperty("os.name").toLowerCase().contains("win")) 
+        {
+            builder = new ProcessBuilder("cmd.exe", "/c", Neodymium.configuration().lighthouseBinaryPath(), "--chrome-flags=\"--ignore-certificate-errors\"", URL, "--port=" + Neodymium.getRemoteDebuggingPort(), "--preset=desktop", "--output=json", "--output=html", "--output-path=target/" + reportName + ".json");
+        }
+        else if (System.getProperty("os.name").toLowerCase().contains("linux") || System.getProperty("os.name").toLowerCase().contains("mac"))
+        {
+            builder = new ProcessBuilder(Neodymium.configuration().lighthouseBinaryPath(), "--chrome-flags=\"--ignore-certificate-errors\"", URL, "--port=" + Neodymium.getRemoteDebuggingPort(), "--preset=desktop", "--output=json", "--output=html", "--output-path=target/" + reportName + ".json");
```
Comment (wurzelkuchen) [inline src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java:157]: for linux please change this to call the shell by using `new ProcessBuilder("sh", "-c", Neodymium.config....`

File: src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+                int errorCode = p.waitFor();
+                
+                if (errorCode != 0) 
+                {
+                    throw new IOException();
+                }
+            }
+            else if (System.getProperty("os.name").toLowerCase().contains("linux") || System.getProperty("os.name").toLowerCase().contains("mac"))
+            {
+                new ProcessBuilder(Neodymium.configuration().lighthouseBinaryPath(), "--version").start();
```
Comment (wurzelkuchen) [inline src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java:57]: for linux please change this to call the shell by using `new ProcessBuilder("sh", "-c", Neodymium.config....`

**Review by wurzelkuchen — CHANGES_REQUESTED** [review]
nice work, now for the polishing ;)

File: config/neodymium.properties [hunk 5]
```diff
@@ -226,6 +226,18 @@ neodymium.webDriver.keepBrowserOpenOnFailure = false
 # If false: all tests of a test class are executed
 neodymium.workInProgress = false
 
+#############################
+#
+# Lighthouse
+#
+#############################
+neodymium.lighthouse.binaryPath = lighthouse
```
[the remark this answers is not in the document]
Comment (kqmpetenz) [reply config/neodymium.properties:256]: done

File: config/neodymium.properties [hunk 6]
```diff
@@ -226,6 +226,18 @@ neodymium.webDriver.keepBrowserOpenOnFailure = false
...
+neodymium.lighthouse.assert.thresholdScore.performance = 0.5
```
[the remark this answers is not in the document]
Comment (kqmpetenz) [reply config/neodymium.properties:257]: done

File: config/neodymium.properties [hunk 7]
```diff
@@ -226,6 +226,18 @@ neodymium.webDriver.keepBrowserOpenOnFailure = false
...
+neodymium.lighthouse.assert.thresholdScore.accessibility = 0.5
+neodymium.lighthouse.assert.thresholdScore.bestPractices = 0.5
+neodymium.lighthouse.assert.thresholdScore.seo = 0.5
+#neodymium.lighthouse.assert.audits =
```
[the remark this answers is not in the document]
Comment (kqmpetenz) [reply config/neodymium.properties:261]: documentation is still todo

File: src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.jayway.jsonpath.JsonPath;
+import com.jayway.jsonpath.PathNotFoundException;
+
+import io.qameta.allure.Allure;
+
+/**
+ * Powered by <a href="https://developer.chrome.com/docs/lighthouse/overview?hl=de">Lighthouse</a> (Copyright Google)
```
[the remark this answers is not in the document]
Comment (kqmpetenz) [reply src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java:24]: done

File: src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+        // get report json scores
+        JsonObject categories = json.getAsJsonObject("categories");
+        double performanceScore = categories.getAsJsonObject("performance").get("score").getAsDouble();
+        double accessibilityScore = categories.getAsJsonObject("accessibility").get("score").getAsDouble();
+        double bestPracticesScore = categories.getAsJsonObject("best-practices").get("score").getAsDouble();
+        double seoScore = categories.getAsJsonObject("seo").get("score").getAsDouble();
+        
+        // validate if values in report json are greater than defined threshold in config
+        SelenideAddons.wrapAssertionError(() -> {
+            Assert.assertTrue("the performance score " + performanceScore + " doesn't exceed nor match the required threshold of " + Neodymium.configuration().lighthouseAssertPerformance() + ", please improve the score to match expectations", Neodymium.configuration().lighthouseAssertPerformance() <= performanceScore);
```
[the remark this answers is not in the document]
Comment (kqmpetenz) [reply src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java:97]: 1. done
2. done

File: src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+            Assert.assertTrue("the best practices score " + bestPracticesScore + " doesn't exceed nor match the required threshold of " + Neodymium.configuration().lighthouseAssertBestPractices() + ", please improve the score to match expectations", Neodymium.configuration().lighthouseAssertBestPractices() <= bestPracticesScore);
+            Assert.assertTrue("the seo score " + seoScore + " doesn't exceed nor match the required threshold of " + Neodymium.configuration().lighthouseAssertSeo() + ", please improve the score to match expectations", Neodymium.configuration().lighthouseAssertSeo() <= seoScore);
+        });
+        
+        // validate jsonpaths in neodymium properties
+        validateAudits(jsonFile);
+        
+        // add report html to allure
+        Allure.addAttachment(reportName, "text/html", FileUtils.openInputStream(new File("target/" + reportName + ".report.html")), "html");
```
[the remark this answers is not in the document]
Comment (kqmpetenz) [reply src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java:107]: done

File: src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+                }
+                catch (PathNotFoundException e)
+                {
+                    continue;
+                }
+                
+            }
+            if (errorAudits.size() > 0) 
+            {
+                throw new Exception("the following audits " + errorAudits + " contain errors that need to be fixed, please look into the corresponding html for further information");
```
[the remark this answers is not in the document]
Comment (kqmpetenz) [reply src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java:213]: nope, done

[the remark this answers is not in the document]
Comment (kqmpetenz) [reply src/main/java/com/xceptance/neodymium/util/LighthouseUtils.java:213]: done

**Review by kqmpetenz — COMMENTED** [review]
only the documentation of the audits property and the link to the documentation in the neo config (when finished)

File: config/neodymium.properties (hunk 7 above)
[the remark this answers is not in the document]
Comment (kqmpetenz) [reply config/neodymium.properties:288]: done

**Review by wurzelkuchen — CHANGES_REQUESTED** [review]

**Review by wurzelkuchen — APPROVED** [review]

**Pull request merged**